### PR TITLE
Add non-indexed fields to ecs templates

### PR DESCRIPTION
**Pull request by flash1293**
Some fields in ECS are specified to not get indexed. The existing default component template `ecs@mappings` isn't capturing this, so the fields will fall back to regular `keyword` mapping:
https://www.elastic.co/guide/en/ecs/current/ecs-event.html#field-event-original

Aligned with https://github.com/elastic/elastic-package/pull/1733 - I'm not sure about the x509 fields, could you elaborate on these @jsoriano ?

This PR fixes the problem by adding dynamic templates for these fields.

### CODE REVIEW

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 1]
```diff
@@ -23,6 +23,26 @@
...
           }
         },
+        {
+          "ecs_event_original_non_indexed_keyword": {
+            "mapping": {
+              "type": "keyword",
+              "index": false,
+              "doc_values": false
+            },
+            "path_match": "event.original"
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:33]: Can we make this an array like we have here? https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/template-resources/src/main/resources/ecs%40mappings.json#L19 Like this the overall template stays simpler and is easy to extend.

**Comment by ruflin** [comment]
@eyalkoren Would be great to get your look at this one if you find some time. I wonder why this did not pop up in the tests and how we can adjust the tests to cover this use case.

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 2]
```diff
@@ -23,6 +23,26 @@
...
+          }
+        },
+        {
+          "ecs_x509_public_key_exponent_non_indexed_keyword": {
+            "mapping": {
+              "type": "keyword",
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:39]: it seems this is a long according to ECS? https://www.elastic.co/guide/en/ecs/8.11/ecs-x509.html#field-x509-public-key-exponent I really wonder if we should cover this "edge" case.

Comment (flash1293) [reply x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json]: I think I agree. ECS doesn't mention whether it's indexed or not - so I guess indexing is fair game (and it's left to the user to overwrite this)?

**Comment by jsoriano** [comment]
> I'm not sure about the x509 fields, could you elaborate on these @jsoriano ?

They are defined as non-indexed in ECS: https://github.com/elastic/ecs/blob/ee4e0979dbaa91c915a1b31a26e1ea814bfe75aa/schemas/x509.yml#L203

**Comment by eyalkoren** [comment]
@ruflin the tests currently ignore the [`index` property](https://github.com/elastic/ecs/blob/ee4e0979dbaa91c915a1b31a26e1ea814bfe75aa/generated/ecs/ecs_flat.yml#L3361) of the ECS field definitions. Since they are defined in the schema, it should be easy to adjust them. I'll try to get to that this week, and bring into consideration the [`doc_values` property](https://github.com/elastic/ecs/blob/ee4e0979dbaa91c915a1b31a26e1ea814bfe75aa/generated/ecs/ecs_flat.yml#L3357) as well.

**Comment by flash1293** [comment]
Thanks for the context @jsoriano - it threw me off that `event.original` is specifying this explicitly in the user-visible text in the docs.

**Comment by flash1293** [comment]
@elasticsearchmachine run

**Comment by flash1293** [comment]
@elasticmachine run

**Comment by ruflin** [comment]
* @eyalkoren That would be great. Should we wait with this PR for the change as we might unconver other things?
* @eyalkoren @leehinman Do we need to increase the version of the template or anything if we update it?
* @flash1293 Try to merge in main, the failures seem to be unrelated

**Comment by ruflin** [comment]
As usual, I pinged the wrong @leehinman , now added @dakrone .

**Comment by eyalkoren** [comment]
> @eyalkoren That would be great. Should we wait with this PR for the change as we might unconver other things?

I don't think we need to wait with the test. My intention was to do it in a separate PR anyway to see it failing and then verify it gets fixed after merging this PR's fix. And I am not just saying that because my availability is non problematic 😬 

 > @eyalkoren @dakrone Do we need to increase the version of the template or anything if we update it?

Since this template uses a [version variable](https://github.com/elastic/elasticsearch/blob/b4938e16457dc69d392235eaf404a6dad9ddb717/x-pack/plugin/core/template-resources/src/main/resources/ecs%40mappings.json#L202) which is used by all other components managed by the stack registry, I believe we would need to update the [registry version](https://github.com/elastic/elasticsearch/blob/b4938e16457dc69d392235eaf404a6dad9ddb717/x-pack/plugin/stack/src/main/java/org/elasticsearch/xpack/stack/StackTemplateRegistry.java#L50), unless it was already updated after the latest release.

**Comment by eyalkoren** [comment]
I ended up pushing the ECS test adjustment here and verified that it fails when removing the `ecs@mappings` fixes.
Since I also contributed, I hereby approve the mappings fix but will let other reviewers to "officially" approve the entire PR changes.

**Review by felixbarny — COMMENTED** [review]
Please increment `org.elasticsearch.xpack.stack.StackTemplateRegistry#REGISTRY_VERSION`, otherwise LGTM

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 3]
```diff
@@ -23,6 +23,30 @@
...
+              "index": false,
+              "doc_values": false
+            },
+            "path_match": [
+              "event.original"
+            ]
+          }
+        },
+        {
+          "ecs_x509_public_key_exponent_non_indexed_long": {
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:39]:
```suggestion
          "ecs_non_indexed_long": {
```

Reason is that we can eventually extend it with other fields. Also changing it later should not be a breaking change.

File: x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json [hunk 4]
```diff
@@ -23,6 +23,30 @@
             "unmatch_mapping_type": "object"
           }
         },
+        {
+          "ecs_event_original_non_indexed_keyword": {
```
Comment (ruflin) [inline x-pack/plugin/core/template-resources/src/main/resources/ecs@mappings.json:27]:
```suggestion
          "ecs_non_indexed_keyword": {
```

See comment below

**Review by ruflin — APPROVED** [review]

**Review by felixbarny — APPROVED** [review]

**Review by jsoriano — APPROVED** [review]

**Review by dakrone — APPROVED** [review]
LGTM

**Pull request merged**